core/modules/wmgr/python/auth.py

```diff
@@ -0,0 +1,121 @@
+
+@author Andy Salnikov - salnikov@slac.stanford.edu
+"""
+
+#--------------------------------
+#  Imports of standard modules --
+#--------------------------------
+import logging
+
+#-----------------------------
+# Imports for other modules --
+#-----------------------------
+from flask import request
+from .flask_httpauth import HTTPBasicAuth, HTTPDigestAuth
+
+#----------------------------------
+# Local non-exported definitions --
+#----------------------------------
+
+_log = logging.getLogger('auth')
+
+#------------------------
+# Exported definitions --
+#------------------------
+
+class Auth(object):
+    """
+    Main responsibility of this class is to verify that user is authenticated
+    with wmgr. We support two different authentication mechanisms, basic or
+    digest, selected via application configuration option AUTH_TYPE.
+
+    Code that needs to check authentication will later call:
+
+        auth = Auth(app.config)
+        response = auth.checkAuth()
+
+    which returns None if authentication suceeds or response object if it
+    fails. Primary use of this method is as an argument to
+    app.before_request() method.
+    """
+
+    def __init__(self, appConfig):
+        """
+        Make new instance of the class.
+
+        @param appConfig:    application configuration instance
+        """
+
+        authType = appConfig.get('AUTH_TYPE', 'digest').lower()
+        _log.debug('wmgr auth type: %s', authType)
+        if authType not in ['none', 'basic', 'digest']:
+            raise RuntimeError("unexpected auth type in configuration: " + authType)
+
+        if authType == "none":
+
+            self._secret = None
+
+        else:
+
+            secretFile = appConfig.get('SECRET_FILE')
+            if secretFile is None:
+                raise RuntimeError("secret file location is required")
+
+            # will fail if secret file is missing
+            self._secret = self.readSecret(secretFile)
+
+            if authType == 'basic':
+                self._auth = HTTPBasicAuth()
+            else:
+                # digest
+                self._auth = HTTPDigestAuth()
+
+    def checkAuth(self):
+        """
+        Check for successful authentication. This method is supposed
+        to be used by app in before_request list, it returns response
+        object if it fails to authenticate or None otherwise.
+        """
+
+        if self._secret:
+            password = None
+            auth = request.authorization
+            if auth and auth.username == self._secret[0]:
+                password = self._secret[1]
+            if not self._auth.authenticate(auth, password):
+                return self._auth.auth_error_callback()
+
+    @staticmethod
+    def readSecret(fileName):
+        """
+        Reads secret file and returns (user, password) tuple
+        """
+        _log.debug('reading wmgr secret from %s', fileName)
+        try:
+            return open(fileName).read().strip().split(':', 1)
```

Comment:
If this is an empty file or the file doesn't contain `:`, this returns a list with only one entry. That would definitely break line 108 but only at runtime. Maybe not a big deal.